src/pudl/analysis/plant_parts_eia.py

```diff
@@ -1415,6 +1506,55 @@ def execute(self, part_df, true_grans):
         return part_df
 
 
+class TrueGranLabeler:
+    """Label the plant-part table records with their true granularity."""
+
+    def execute(self, ppl):
+        """Merge the true granularity labels onto the plant part df.
+
+        First the plant part list records are matched to generators. Then
+        the matched records are sorted by PLANT_PARTS_ORDERED and the
+        highest granularity record for each generator is marked as the true
+        granularity. The appropriate true granular part label and record id
+        is then merged on to get the plant part table with true granularity labels.
+
+        Arguments:
+            ppl: (pd.DataFrame) The plant parts list
+        """
+        # What type of merge do we want to do in the match_to_single_part? Outer merge?
+        # i think something is wrong because of ownership dupes in gens_mega?
+        parts_to_gens = PlantPart(part_name="plant_gen").match_to_single_plant_part(
+            multi_gran_df=ppl, ppl=ppl, cols_to_keep=["plant_part"]
+        )
+        # categorical columns allow sorting by PLANT_PARTS_ORDERED
+        parts_to_gens["plant_part_og"] = pd.Categorical(
+            parts_to_gens["plant_part_og"], PLANT_PARTS_ORDERED
+        )
+        ppl_true_gran = parts_to_gens.sort_values("plant_part_og")
+        # find duplicate generators and mark all as duplicates except the
+        # highest granularity record according to PLANT_PARTS_ORDERED
+        dupes = ppl_true_gran.duplicated(subset=["record_id_eia"], keep="first")
```

Comment:
I thiiiiink that somehow we need to know about the collection of generators that are matched to each plant_part record, not the individual records.